### ISSUE TITLE
wait_for with port & search_regex can run forever

### ISSUE THREAD

**Issue by jan-grant** [ansible/ansible-modules-core#2051]
##### Summary:

wait_for won't always honour its timeout values. In particular, if you connect to a socket which dribbles data out at a rate that exceeds the specified timeout, you'll wind up in a loop reading data forever.

The problem lies with this inner loop in the wait_for code:

```
elif port:
    ...
             while 1:
                 data += s.recv(1024)
                 if not data:
                     break
                 elif re.search(search_regex, data, re.MULTILINE):
                     matched = True
                     break
    ...
```

There may be additional issues around the close timeout in the same body of code; I found that killing the nc side of the connection left the ansible playbook permanently in close_wait. This is down to not checking the result of s.recv directly for an EOF: the whole thing should be as follows:

```
elif port:
    ...
             while datetime.datetime.now() < end:
                 response = s.recv(1024)
                 if not response:
                     break
                 data += response
                 elif re.search(search_regex, data, re.MULTILINE):
                     matched = True
                     break
    ...
```
##### Steps To Reproduce:

An example ansible task:

```
- hosts: localhost
  connection: local
  tasks:
    - name: test wait_for
      wait_for:
        timeout: 5
        delay: 1
        connect_timeout: 5
        host: localhost
        port: 9500
        search_regex: 'is synced'
        state: present
```

Run this with the following running in another window:

```
(while : ; do echo n; sleep 2; done) | nc -l 9500
```
##### Expected Results:

A failure after 5 seconds (in the example above).
##### Actual Results:

The playbook will hang indefinitely; either reading data or, should you kill the server process (nc in the example above), stuck in CLOSE_WAIT and looping forever.

**Comment by gregdek** [comment]
cc: @jhoekx @jarv @AnderEnder

**Comment by abadger** [comment]
@jan-grant or anyone else who's interested in the wait_for module, could you take a look at PR #2373 to see if it does the right thing here?  I don't do very much socket programming so I'd love someone else to test.

It uses select instead of the extension to the existing socket-module-based code to address potential cornercases with connect_timeout and timeout.  For instance, if we ended up waiting nearly the full timeout value before our connection was established, we could end up waiting a second full timeout value waiting for data to be sent from the server.  Using select should fix that.

**Comment by abadger** [comment]
Merged #2373.  Feel free to open a new issue/PR if there's still issues with the new code.

# Closing This Ticket

Hi!

We believe recent commits (likely detailed above) should resolve this question or problem for you.

This will also be included in the next major release.

If you continue seeing any problems related to this issue, or if you have any further questions, please let us know by stopping by one of the two mailing lists, as appropriate:
- https://groups.google.com/forum/#!forum/ansible-project - for user questions, tips, and tricks
- https://groups.google.com/forum/#!forum/ansible-devel - for strategy, future planning, and questions about writing code

Because this project is very active, we're unlikely to see comments made on closed tickets, but the mailing list is a great way to ask questions, or post if you don't think this particular
issue is resolved.

Thank you!
